### ISSUE TITLE
Consider checking that static fields of dynamic arguments are hashable

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#7826]
The following code
```
from jax import jit
from flax.struct import dataclass, field
import numpy as np

@dataclass
class Model():
    c: np.ndarray = field(pytree_node=False)

@jit
def f(model):
    return 1.0

m = Model(np.ones(10))
n = Model(np.ones(10))

f(m)
f(n)
```
Gives a strange error when the jit tries to look up using array-valued static fields
```
Traceback (most recent call last):
  File "/home/neil/a.py", line 17, in <module>
    f(n)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
It might be worth improving the check in `jax/api_util::argnums_partial_except` and `argnames_partial_except` to also check that the static fields of dynamic arguments are hashable.

**Comment by mattjj** [comment]
Great idea!

I really appreciate you opening this issue, and always helping to make JAX better. These improvements really add up!

**Comment by froystig** [comment]
> It might be worth improving the check in `jax/api_util::argnums_partial_except` and `argnames_partial_except` to also check that the static fields of dynamic arguments are hashable.

@NeilGirdhar, is this something that you would be interested in contributing?

**Comment by NeilGirdhar** [comment]
@froystig Okay, I just got unblocked on my research, but I should be able to find some time for this soon.  Also looking forward to the review of https://github.com/google/jax/pull/7158 😄

**Comment by froystig** [comment]
Up to you! Someone on the team will be happy to handle this otherwise (possibly at lower priority). I thought to ask in response to your initial comment, since you seem to have a good handle on the fix.
